File: examples/indirect_writes.f

      program indirect_writes

      include 'lplib3.ins'
      include 'libmeshb7.ins'

      integer*4 MaxVer,MaxTet,NmbItr
      parameter (MaxVer=10000000)
      parameter (MaxTet=60000000)
      parameter (NmbItr=100)

      integer*4 i,j,k,NmbCpu,ver,dim,ref,TetTyp,VerTyp,ret
      integer*4 VerDeg(MaxVer),TetVer(4,MaxTet)
      integer*4 NmbPth,DepTab(4)
      integer*8 NmbVer,NmbTet,ParIdx,InpMsh
      real*8 VerTem(MaxVer),TetTem(MaxTet),t
      real*4 sta(2),acc

      external vertemp,tettemp

      InpMsh = gmfopenmesh("tet.meshb", GmfRead,ver,dim)
      if(InpMsh.eq.0) STOP ' cannot open file tet.meshb'

      NmbVer = gmfstatkwd(InpMsh, GmfVertices)
      if(NmbVer.gt.MaxVer) STOP ' too many vertices'

      NmbTet = gmfstatkwd(InpMsh, GmfTetrahedra)
      if(NmbTet.gt.MaxTet) STOP ' too many tetrahedra'

      print*, 'version  = ',ver,   'dimension  = ',dim
      print*, 'vertices = ',NmbVer,'tetrahedra = ',NmbTet

      ret = gmfgetblock(InpMsh,GmfTetrahedra,1,NmbVer,0_8,
     +                  GmfInt, TetVer(1,1), TetVer(1,NmbVer),
     +                  GmfInt, TetVer(2,1), TetVer(2,NmbVer),
     +                  GmfInt, TetVer(3,1), TetVer(3,NmbVer),
     +                  GmfInt, TetVer(4,1), TetVer(4,NmbVer),
     +                  GmfInt, ref, ref)

      ret = gmfclosemesh(InpMsh)

      ParIdx = initparallel(0)
      if(ParIdx.eq.0) STOP ' cannot init LPLIB3'
      NmbPth = getnumberofcores()
      print*, 'lib = ', ParIdx, 'threads = ', NmbPth

      VerTyp = newtype(ParIdx, NmbVer)
      if(VerTyp.eq.0) STOP ' cannot define vertex type'

      TetTyp = newtype(ParIdx, NmbTet)
      if(TetTyp.eq.0) STOP ' cannot define tetrahedron type'

      print*, 'VerTyp = ', VerTyp, 'TetTyp = ', TetTyp

      do i=1,NmbVer
          VerDeg(i) = 0
          VerTem(i) = i
      end do

      ret = begindependency(ParIdx, TetTyp, VerTyp)
      do i=1,NmbTet
          do j=1,4
              ret = adddependency(ParIdx, i, TetVer(j,i))
              VerDeg(TetVer(j,i)) = VerDeg(TetVer(j,i)) + 1
          end do
      end do
      ret = enddependency(ParIdx, sta)
      print*, 'mean = ', sta(1), 'max = ', sta(2)

      t = getwallclock()

      do i=1,NmbItr
          acc = acc + launchparallel(ParIdx,TetTyp,VerTyp,tettemp,3,
     +                               TetVer,TetTem,VerTem)
           acc = acc + launchparallel(ParIdx,TetTyp,VerTyp,vertemp,4,
     +                               TetTem,VerTem,VerDeg,TetVer)
      end do

      print*, 'wall time = ', getwallclock() - t, 'accelerate = ',
     +         acc / (NmbItr*2)

      call stopparallel(ParIdx)

      end


      subroutine tettemp(BegIdx,EndIdx,PthIdx,TetVer,TetTem,VerTem)

      integer i,j,BegIdx,EndIdx,PthIdx,TetVer(4,*)
      real*8 sum,TetTem(*),VerTem(*)

      do i = BegIdx,EndIdx
          sum = 0
          do j=1,4
              sum = sum + VerTem(TetVer(j,i))
          end do
          TetTem(i) = sum
      end do

      return
      end


      subroutine vertemp(BegIdx,EndIdx,PthIdx,TetTem,
     +                   VerTem,VerDeg,TetVer)

      integer i,j,k,BegIdx,EndIdx,PthIdx,TetVer(4,*),VerDeg(*)
      real*8 TetTem(*),VerTem(*)

      do i = BegIdx,EndIdx
          do j=1,4
              k = TetVer(j,i)
              VerTem(k) = VerTem(k) + TetTem(i) / VerDeg(k)
          end do
      end do

      return
      end

File: examples/indirect_writes.c


/*----------------------------------------------------------------------------*/
/*                                                                            */
/*                PARALLEL INDIRECT MEMORY WRITES USING LPlib                 */
/*                                                                            */
/*----------------------------------------------------------------------------*/
/*                                                                            */
/*   Description:       direct and indirect memory writes                     */
/*   Author:            Loic MARECHAL                                         */
/*   Creation date:     jan 15 2015                                           */
/*   Last modification: feb 01 2017                                           */
/*                                                                            */
/*----------------------------------------------------------------------------*/


/*----------------------------------------------------------------------------*/
/* Includes                                                                   */
/*----------------------------------------------------------------------------*/

#include <stdio.h>
#include <stdlib.h>
#include <float.h>
#include "lplib3.h"
#include "libmeshb7.h"


/*----------------------------------------------------------------------------*/
/* Structure's prototype                                                      */
/*----------------------------------------------------------------------------*/

typedef struct
{
   int TetTyp, VerTyp, NmbVer, NmbTet, (*TetVer)[4], *VerDeg;
   long long ParIdx;
   double *VerTem, *TetTem;
}MshSct;


/*----------------------------------------------------------------------------*/
/* Procedure called in parallel adding the vertices'value                     */
/*  to the tetrahedron's value (INDIRECT READ)                                */
/* Arguments are imposed by the library:                                      */
/* BegIdx: begining loop index                                                */
/* EndIdx: ending loop index                                                  */
/* PthIdx: index of the thread executing this instance                        */
/* arg   : pointer to a user's structure containing every                     */
/*        arguments needed by the loop                                        */
/*----------------------------------------------------------------------------*/

void TetTem(int BegIdx, int EndIdx, int PthIdx, MshSct *msh)
{
   int i, j;
   double t;

   for(i=BegIdx;i<=EndIdx;i++)
   {
      t = 0;

      for(j=0;j<4;j++)
         t += msh->VerTem[ msh->TetVer[i][j] ];

      msh->TetTem[i] = t;
   }
}


/*----------------------------------------------------------------------------*/
/* Procedure called in parallel adding a tetrahedron's                        */
/*  temperature to the vertices' values (INDIRECT WRITE)                      */
/* Arguments are imposed by the library:                                      */
/* BegIdx: begining loop index                                                */
/* EndIdx: ending loop index                                                  */
/* PthIdx: index of the thread executing this instance                        */
/* arg   : pointer to a user's structure containing every                     */
/*        arguments needed by the loop                                        */
/*----------------------------------------------------------------------------*/

void VerTem(int BegIdx, int EndIdx, int PthIdx, MshSct *msh)
{
   int i, j;

   for(i=BegIdx;i<=EndIdx;i++)
      for(j=0;j<4;j++)
         msh->VerTem[ msh->TetVer[i][j] ] += \
            msh->TetTem[i] / msh->VerDeg[ msh->TetVer[i][j] ];
}


/*----------------------------------------------------------------------------*/
/* Setup LPlib and launch 100 temperature soomthing steps                     */
/*----------------------------------------------------------------------------*/

int main(int ArgCnt, char **ArgVec)
{
   int i, j, NmbCpu=GetNumberOfCores(), ver, dim, ref, NmbItr=100;
   long long InpMsh;
   float sta[2], acc=0;
   double tim=0;
   MshSct msh;

   // Read the number of threads to launch from the command line argument
   if(ArgCnt > 1)
      NmbCpu = atoi(*++ArgVec);

   // Open the input mesh
   if(!(InpMsh = GmfOpenMesh("../sample_meshes/tet.meshb", GmfRead, &ver, &dim)))
      return(1);

   puts("");
   printf("Input mesh : idx = %lld, version = %d, dimension = %d\n", \
            InpMsh, ver, dim);

   msh.NmbVer = GmfStatKwd(InpMsh, GmfVertices);
   printf("Input mesh : nmb vertices = %d\n", msh.NmbVer);

   msh.NmbTet = GmfStatKwd(InpMsh, GmfTetrahedra);
   printf("Input mesh : nmb tets = %d\n", msh.NmbTet);

   // Allocate the memory
   msh.TetVer = malloc( (msh.NmbTet+1) * 4 * sizeof(int) );
   msh.TetTem = malloc( (msh.NmbTet+1) * sizeof(double) );
   msh.VerTem = malloc( (msh.NmbVer+1) * sizeof(double) );
   msh.VerDeg = calloc( (msh.NmbVer+1), sizeof(int) );

   // Read the mesh
   GmfGotoKwd(InpMsh, GmfTetrahedra);

   for(i=1;i<=msh.NmbTet;i++)
      GmfGetLin(  InpMsh, GmfTetrahedra, &msh.TetVer[i][0], &msh.TetVer[i][1], \
                  &msh.TetVer[i][2], &msh.TetVer[i][3], &ref );

   GmfCloseMesh(InpMsh);

   // Initialize the LPlib and setup the data types
   if(!(msh.ParIdx = InitParallel(NmbCpu)))
   {
      puts("Error initializing the LPlib.");
      exit(1);
   }

   if(!(msh.TetTyp = NewType(msh.ParIdx, msh.NmbTet)))
   {
      puts("Error while creating tetrahedra data type.");
      exit(1);
   }

   if(!(msh.VerTyp = NewType(msh.ParIdx, msh.NmbVer)))
   {
      puts("Error while creating vertices data type.");
      exit(1);
   }

   puts("");
   printf("TetTyp = %d, VerTyp = %d, NmbCpu = %d\n", \
         msh.TetTyp, msh.VerTyp, NmbCpu);

   /* Setup dependencies between tets and vertices 
      and compute vertices' degree on the flight */

   BeginDependency(msh.ParIdx, msh.TetTyp, msh.VerTyp);

   for(i=1;i<=msh.NmbTet;i++)
      for(j=0;j<4;j++)
      {
         AddDependency(msh.ParIdx, i, msh.TetVer[i][j]);
         msh.VerDeg[ msh.TetVer[i][j] ]++;
      }

   EndDependency(msh.ParIdx, sta);

   printf("dependencies stats : average = %g %%, maximum = %g %%\n", \
            sta[0], sta[1]);

   // Initialize the vertices' temperature with some crap values
   for(i=1;i<=msh.NmbVer;i++)
      msh.VerTem[i] = random();

   // Perform temperature smoothing steps
   tim = GetWallClock();

   puts("");
   for(i=1;i<=NmbItr;i++)
   {
      if(!(acc += LaunchParallel(msh.ParIdx, msh.TetTyp, msh.VerTyp, \
                                 (void *)TetTem, (void *)&msh)))
      {
         puts("Error while launching the parallel loop TetTem.");
         exit(1);
      }

      if(!(acc += LaunchParallel(msh.ParIdx, msh.TetTyp, 0, \
                                 (void *)VerTem, (void *)&msh)))
      {
         puts("Error while launching the parallel loop VerTem.");
         exit(1);
      }
   }

   tim = GetWallClock() - tim;

   printf(" %d steps, average concurency = %g, // running time = %gs\n", \
         NmbItr, acc / (2 * NmbItr), tim);

   // Stop and free everything
   StopParallel(msh.ParIdx);

   free(msh.TetVer);
   free(msh.TetTem);
   free(msh.VerTem);
   free(msh.VerDeg);

   return(0);
}
